README.md
<p>This project is a <b>python module</b> to worik with <b>NooLite USB stick (PC118, PC1116, PC1132)</b>.<br>
Can be easy used to make a light control for smarthomes.<br>
About NooLite: http://www.noo.com.by/</p>


<p>Look at wiki for more info.</p>

<p>Author: Anton Balashov<br>
E-mail: sicness(_at_)darklogic.ru<br>
License: GPL v3<br>
Site: https://github.com/Sicness/pyNooLite</p>

<h4>Install</h4>

    sudo apt-get install pip
    sudo pip install noolite

<h3>Ubuntu/debian</h3
Ubuntu and debian users can install the module with an executable example from ppa:<br>
    https://launchpad.net/~ctolbhuk/+archive/noolite

<b>Or </b>download this repo and go into.

    sudo python setup.py install

<b>Or </b>simple copy noolite.py from this repo to your project<br>

<H4>Example:</H4>
    import noolite

    n = noolite.NooLite()
    n.on(0)       # Turn power on on first channel
    n.off(0)      # Turn power off on first channel
    n.set(1, 115) # Set 115 level on second channel
    n.bind(7)     # send "bind" signal on channel 8

<p>See noolite file for a extra example</p>

<h4>Dependences:</h4>

*  python module pyusb<br>
For Ubuntu do: <br>
     <i>sudo apt-get install python-pip <br>
     sudo pip install pyusb</i><br>

* To have access on device from common user add the next rule to udev. For example to /etc/udev/rules.d/50-noolite.rules next line:<br>
ATTRS{idVendor}=="16c0", ATTRS{idProduct}=="05df", SUBSYSTEMS=="usb", ACTION=="add", MODE="0666", GROUP="dialout"<br>
And add your user to dialout group:<br>
    <i>sudo usermod &lt;user&gt; -a -G dialout</i>


CHANGES.txt
noolite (1.2.2)
  * Fixed 'save scenario' command
  * Fixed 'on', 'off' and other simple commands sending incorrect format and level data after 'set' command calls
  * Deprecated 'channals' NooLite class init argument in favour of correctly spelled 'channels' one
  * Deprecated 'tests' NooLite class init argument
  * TypeErrors raised instead of ValueErrors where appropriate
  * General code refactoring
  * PEP8 corrections
  * Corrections of error messages and descriptions
  * Added .gitignore file to github repository

 -- Yaraslau Zhylko <yaraslauzhylko@gmail.com> Thu, 24 Mar 2016 18:30:00 +0400

noolite (1.2.1)
  * Fixed unbind(self, ch) function with correct 0x09 command
  * Added option to refine USB device search by passing optional pyusb core.find() arguments to NooLite object constructor
  * Specialised NooLiteDeviceLookupErr is now raised if device is not found / not connected

 -- Yaraslau Zhylko <yaraslauzhylko@gmail.com> Sun, 13 Mar 2016 11:33:33 +0400

noolite (1.1.3)
  * Impoved tests. Now tests checks returned cmd
  * New tests for set function

 -- Anton Balashov <sicness@darklogic.ru> Thu, 25 Jun 2013 09:24:33 +0400

noolite (1.1.2)
  * Fix set level function
  * Some PEP8 work

 -- Anton Balashov <sicness@darklogic.ru> Thu, 20 Jun 2013 13:35:35 +0400

noolite (1.1.1)
  * Added try on no or old pyusb in example

 -- Anton Balashov <sicness@darklogic.ru> Thu, 23 May 2013 22:15:40 +0400

noolite (1.1.0)

  * PEP8 for noolite.py and noolite
  * check import pyusb and print info on error

 -- Anton Balashov <sicness@darklogic.ru> Thu, 21 May 2013 12:50:00 +0400


tests.py
import unittest

try:
    from unittest import mock
except ImportError:
    import mock

import noolite

init_cmd = [0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]


class Tests(unittest.TestCase):

    def setUp(self):
        self.exp_cmd = init_cmd[:]

    def test_init_ch_type(self):
        self.assertRaises(
            ValueError, noolite.NooLite, channals="bla", tests=True)

    def test_init_idVendor_type(self):
        self.assertRaises(
            ValueError, noolite.NooLite, idVendor="bla", tests=True)

    def test_init_idProduct_type(self):
        self.assertRaises(
            ValueError, noolite.NooLite, idProduct="bla", tests=True)

    def test_ch_str(self):
        switcher = noolite.NooLite(tests=True)
        self.exp_cmd[4] = 7
        self.exp_cmd[1] = 0x02
        self.assertEqual(switcher.on("7"), self.exp_cmd)

    def test_ch_int(self):
        switcher = noolite.NooLite(tests=True)
        self.exp_cmd[4] = 7
        self.exp_cmd[1] = 0x00
        self.assertEqual(switcher.off(7), self.exp_cmd)

    def test_ch_negative(self):
        switcher = noolite.NooLite(tests=True)
        self.assertRaises(noolite.NooLiteErr, switcher.on, -1)

    def test_ch_too_big(self):
        switcher = noolite.NooLite(channals=32, tests=True)
        self.assertRaises(noolite.NooLiteErr, switcher.on, 42)

    def test_on(self):
        switcher = noolite.NooLite(tests=True)
        self.exp_cmd[4] = 7
        self.exp_cmd[1] = 0x02
        self.exp_cmd[2] = 0x00
        self.exp_cmd[5] = 0x00
        self.assertEqual(switcher.on(7), self.exp_cmd)

    def test_off(self):
        switcher = noolite.NooLite(tests=True)
        self.exp_cmd[4] = 7
        self.exp_cmd[1] = 0x00
        self.exp_cmd[2] = 0x00
        self.exp_cmd[5] = 0x00
        self.assertEqual(switcher.off(7), self.exp_cmd)

    def test_set(self):
        switcher = noolite.NooLite(tests=True)
        value = 50
        self.exp_cmd[4] = 7
        self.exp_cmd[2] = 0x01
        self.exp_cmd[1] = 0x06
        self.exp_cmd[5] = 35 + value
        self.assertEqual(switcher.set(7, value), self.exp_cmd)

    def test_set_str(self):
        switcher = noolite.NooLite(tests=True)
        value = "70"
        self.exp_cmd[4] = 7
        self.exp_cmd[2] = 0x01
        self.exp_cmd[1] = 0x06
        self.exp_cmd[5] = 35 + int(value)
        self.assertEqual(switcher.set(7, value), self.exp_cmd)

    def test_set_zero(self):
        switcher = noolite.NooLite(tests=True)
        value = 0
        self.exp_cmd[4] = 7
        self.exp_cmd[2] = 0x01
        self.exp_cmd[1] = 0x06
        self.exp_cmd[5] = 0
        self.assertEqual(switcher.set(7, value), self.exp_cmd)

    def test_switch(self):
        switcher = noolite.NooLite(tests=True)
        self.exp_cmd[4] = 7
        self.exp_cmd[1] = 0x04
        self.exp_cmd[2] = 0x00
        self.exp_cmd[5] = 0x00
        self.assertEqual(switcher.switch(7), self.exp_cmd)

    def test_save(self):
        switcher = noolite.NooLite(tests=True)
        self.exp_cmd[4] = 7
        self.exp_cmd[1] = 0x08
        self.exp_cmd[2] = 0x00
        self.exp_cmd[5] = 0x00
        self.assertEqual(switcher.save(7), self.exp_cmd)

    def test_load(self):
        switcher = noolite.NooLite(tests=True)
        self.exp_cmd[4] = 7
        self.exp_cmd[1] = 0x07
        self.exp_cmd[2] = 0x00
        self.exp_cmd[5] = 0x00
        self.assertEqual(switcher.load(7), self.exp_cmd)

    def test_bind(self):
        switcher = noolite.NooLite(tests=True)
        self.exp_cmd[4] = 7
        self.exp_cmd[1] = 0x0f
        self.exp_cmd[2] = 0x00
        self.exp_cmd[5] = 0x00
        self.assertEqual(switcher.bind(7), self.exp_cmd)

    def test_unbind(self):
        switcher = noolite.NooLite(tests=True)
        self.exp_cmd[4] = 7
        self.exp_cmd[1] = 0x09
        self.exp_cmd[2] = 0x00
        self.exp_cmd[5] = 0x00
        self.assertEqual(switcher.unbind(7), self.exp_cmd)

    def test_simple_cmds_reset_format_and_data_bytes_after_set_cmd_call(self):
        switcher = noolite.NooLite(tests=True)
        level = 67
        for method in ('on', 'off', 'switch', 'bind', 'unbind'):
            # Call set method
            act_cmd = switcher.set(7, level)
            self.assertEqual(act_cmd[1], 0x06)  # Command
            self.assertEqual(act_cmd[2], 0x01)  # Format
            self.assertEqual(act_cmd[5], 35 + level)  # Level
            # Call non-set method
            act_cmd = getattr(switcher, method)(7)
            self.assertNotEqual(act_cmd[1], 0x06)  # Command
            self.assertEqual(act_cmd[2], 0x00)  # Format
            self.assertEqual(act_cmd[5], 0x00)  # Level

    def test_device_kwargs_used_for_device_finding(self):
        switcher = noolite.NooLite(bus=1, address=12)
        with mock.patch('usb.core.find', return_value=mock.Mock()) \
                as mock_usb_core_find:
            switcher.on(5)
        mock_usb_core_find.assert_called_once_with(
            idVendor=mock.ANY, idProduct=mock.ANY,
            bus=1, address=12)

    def test_NooLiteDeviceLookupErr(self):
        switcher = noolite.NooLite(idVendor=12345, idProduct=54321)
        with mock.patch('usb.core.find', return_value=None):
            with self.assertRaises(noolite.NooLiteDeviceLookupErr) as err:
                switcher.off(5)
        self.assertIsInstance(err.exception, noolite.NooLiteErr)
        self.assertIn('idVendor=12345', err.exception.value)
        self.assertIn('idProduct=54321', err.exception.value)

    def test_NooLiteDeviceLookupErr_with_device_kwargs(self):
        switcher = noolite.NooLite(
            idVendor=12345, idProduct=54321, bDeviceClass=7, bDeviceProtocol=1)
        with mock.patch('usb.core.find', return_value=None):
            with self.assertRaises(noolite.NooLiteDeviceLookupErr) as err:
                switcher.off(5)
        self.assertIn('idVendor=12345', err.exception.value)
        self.assertIn('idProduct=54321', err.exception.value)
        self.assertIn('bDeviceClass=7', err.exception.value)
        self.assertIn('bDeviceProtocol=1', err.exception.value)

    def test_NooLiteDeviceLookupErr_with_custom_match_device_kwarg(self):
        switcher = noolite.NooLite(custom_match=lambda dev: dev.bus == 2)
        with mock.patch('usb.core.find', return_value=None):
            with self.assertRaises(noolite.NooLiteDeviceLookupErr) as err:
                switcher.off(5)
        self.assertIn('custom_match=<function', err.exception.value)


if __name__ == '__main__':
    unittest.main(verbosity=2)


main.py
#!/usr/bin/env python

"""
Example of usage python module for NooLite USB stick

Last version at: https://github.com/Sicness/pyNooLite
"""
import sys
import noolite

__author__ = "Anton Balashov"
__license__ = "GPL v3"
__maintainer__ = "Anton Balashov"
__email__ = "Sicness@darklogic.ru"

def help():
    print """
    Example of usage NooLite class

    USAGE:
    on <ch>         Turn power on on channel
    off <ch>        Turn power off on channel
    switch <ch>     Switch between on and off on channel
    set <ch> <lvl>  Set power level on channel
    save <ch>       Save state on channel to scenario
    load <ch>       Call saved scenario
    bind <ch>       Send bind signal on channel
    unbind <ch>     Send unbind signal on channel

    channales should be in 0-7
    level should be in 0-120

    Examples:
        $ python main.py on 1
        $ python main.py set 0 115
    """

if __name__ == '__main__':
    if len(sys.argv) < 3:
        help()
        sys.exit(0)

    cmd = sys.argv[1]
    ch = sys.argv[2]
    if cmd == 'set':
        if len(sys.argv) == 4:
            lvl = int(sys.argv[3])
        else:
            sys.stderr.write("command 'set' must have additional agument 'lvl'")
            help()
            sys.exit(1)

    n = noolite.NooLite()
    cmds = {'on' : n.on,
        'off': n.off,
        'switch': n.switch,
        'set': n.set,
        'save': n.save,
        'load': n.load,
        'bind': n.bind,
        'unbind': n.unbind}

    if not cmd in cmds:
        sys.stderr.write("Unknow command %s\n" % (cmd))
        help()
        sys.exit(4)

    try:
        if sys.argv[1] == 'set':
            cmds[cmd](ch, lvl)
        else:
            cmds[cmd](ch)
    except (ValueError, noolite.NooLiteErr) as e:
        sys.stderr.write(str(e))
        sys.exit(3)


noolite.py
#!/usr/bin/env python

## Python module for working with NooLite USB stick

## Last version at: https://github.com/Sicness/pyNooLite

import sys

try:
    import usb.core
except:
    sys.stderr.write("""
  Can't import usb.core
  please, install pyusb from pypi:
    pip install pyusb
  or
    pip install pyusb --upgrade
  if needed\n\n""")
    raise


__author__ = "Anton Balashov"
__license__ = "GPL v3"
__maintainer__ = "Anton Balashov"
__email__ = "Sicness@darklogic.ru"


class NooLiteErr(Exception):
    def __init__(self, value):
        self.value = value

    def __str__(self):
        return repr(self.value)


class NooLiteDeviceLookupErr(NooLiteErr):
    pass


class NooLite:

    _init_command = [0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]

    def __init__(self, channals=8, idVendor=0x16c0,
                 idProduct=0x05df, tests=False, **device_kwargs):
        if (type(idVendor) != type(int())) or (type(idProduct)) != type(int()):
            raise ValueError("idVendor and idProduct must has int type")
        if type(channals) != type(int()):
            raise ValueError("channals must has int type")
        self._idVendor = idVendor
        self._idProduct = idProduct
        self._device_kwargs = device_kwargs
        self._channales = channals
        self._tests = tests

    def _set_ch(self, ch):
        try:
            ch = int(ch)
        except ValueError:
            raise ValueError("channal has %s can't be converted to int"
                             % (type(ch)))
        if (ch >= self._channales) or (ch < 0):
            raise NooLiteErr("Can't work with %d channal" % (ch))
        self._cmd[4] = ch

    def _send(self):
        if self._tests:    # if it's just a unittests
            return self._cmd
        # find NooLite usb device
        dev = usb.core.find(
            idVendor=self._idVendor, idProduct=self._idProduct,
            **self._device_kwargs)
        if dev is None:
            error = "Can't find device with idVendor=%d, idProduct=%d" \
                % (self._idVendor, self._idProduct)
            if self._device_kwargs:
                error += ", " + ", ".join(
                    ["%s=%s" % (k, v) for k, v in self._device_kwargs.items()])
            raise NooLiteDeviceLookupErr(error)
        if dev.is_kernel_driver_active(0) is True:
            dev.detach_kernel_driver(0)
        dev.set_configuration()
        dev.ctrl_transfer(0x21, 0x09, 0, 0, self._cmd)
        return 0

    def on(self, ch):
        """Turn power on on channel
        First channal is 0 """
        self._cmd = self._init_command[:]
        self._cmd[1] = 0x02       # "Turn power on" command
        if self._set_ch(ch):
            return -2
        return self._send()

    def off(self, ch):
        """Turn power off on channel
        First channal is 0 """
        self._cmd = self._init_command[:]
        self._cmd[1] = 0x00       # "Turn power off" command
        if self._set_ch(ch):
            return -2
        return self._send()

    def set(self, ch, level):
        """Set brightness level
        Max level is 120
        First channal is 0 """
        self._cmd = self._init_command[:]
        self._cmd[1] = 0x06       # Send 'set level' command
        if self._set_ch(ch):
            return -2
        self._cmd[2] = 0x01       # set flag for use 'value'

        # level in cmd must be in [0, 35 - 155]
        try:
            level = int(level)
        except ValueError:
            raise ValueError("level has %s and can't be converted to int"
                             % (type(level)))
        if level == 0:
            self._cmd[5] = 0
        elif level > 120:
            self._cmd[5] = 155
        else:
            self._cmd[5] = 35 + level
        return self._send()

    def save(self, ch):
        """Save state on channel to scenario
        First channal is 0 """
        self._cmd = self._init_command[:]
        self._cmd[1] = 0x08       # "Save scenario" command
        if self._set_ch(ch):
            return -2
        return self._send()

    def load(self, ch):
        """Call saved scenario on channel
        First channal is 0 """
        self._cmd = self._init_command[:]
        self._cmd[1] = 0x07       # "Load scenarion" command
        if self._set_ch(ch):
            return -2
        return self._send()

    def bind(self, ch):
        """ Send bind signal on channel
        First channal is 0 """
        self._cmd = self._init_command[:]
        self._cmd[1] = 0x0f       # "bind" command
        if self._set_ch(ch):
            return -2
        return self._send()

    def unbind(self, ch):
        """ Send unbind signal on channel
        First channal is 0 """
        self._cmd = self._init_command[:]
        self._cmd[1] = 0x09       # "unbind" command
        if self._set_ch(ch):
            return -2
        return self._send()

    def switch(self, ch):
        """switch power between off and on on channel
        First channal is 0 """
        self._cmd = self._init_command[:]
        self._cmd[1] = 0x04       # "switch" command
        if self._set_ch(ch):
            return -2
        return self._send()
